### PR TITLE
fix: getAllowedResources for all namespaces using SelfSubjectRulesReview

### LINKED ISSUE #1700: Lens 4.0 Breaking on MAC
**Describe the bug**
Unable to use & load any section after upgrading to 4.0.0 on MAC.

**To Reproduce**
1. Update to 4.0.0
2. Start Lens
3. Open Developer console to see the error

**Expected behavior**
All sections should load the Kubernetes related resources properly

**Screenshots**
Screen Shot 2020-12-08 at 6 41 30 PM

**Environment (please complete the following information):**
- Lens Version: 4.0.0
- OS: [e.g. MacOS High Sierra Version 10.13.6]
- Installation method (e.g. snap or AppImage in Linux): Auto Upgrade

### LINKED ISSUE #2399: Lens making too many network call even when the UI is closed
**Describe the bug**
We have been seeing huge number of network calls made by lens. I assume it does heavy polling against cluster but the issue is that even when the I close the app, network traffic continues. i.e when app is closed but the lens icon is still visible in mac os system tray at the top

**To Reproduce**
1. Open lens
2. Open activity monitor on mac > network 
3. Notice the network traffic
3. Now close the lens app by clicking on red close button in the lens app ( do not force quite from system tray)
4. Open activity monitor on mac > network 
5. You will notice there is still a lot of network activity.

**Expected behavior**
When the UI is closed the background lens process shouldn't spam cluster

**Screenshots**
Following network activity by lens when ui is not even running
Screen Shot 2021-03-26 at 10 29 45 AM

**Environment (please complete the following information):**
- Lens Version: 4.1.5
- OS: max os x 10.15.6
- DMG

### LINKED ISSUE #2609: In AWS EKS the list of pods is visible from `kubectl` not from Lens.
**Describe the bug**
I'm using EKS on AWS.
If I use `kubectl` I can get the list of running pods, their logs and so on.
In lens I cannot access the pods list instead under workload.

**To Reproduce**
I use this procedure to reproduce the bug:

1. From command line I launch  `aws sso login --profile=ic-test`
2. That completes successfully, opens a browser, complete the login, etc...
3. I open Lens and I check the workload section, nothing there.
4. But when I issue a kubectl get pods from command line, pods are shown

**Expected behavior**
I think I should see the pods in the workload section

**Screenshots**
1) The login to AWS is successful
image

2) Pods are not shown in Workload but I can get them with `kubectl`
image

**Environment (please complete the following information):**
- Lens Version: 4.2.2
- OS: Windows10
- Installation method (e.g. snap or AppImage in Linux): Win exe 64 bit, installation for all users

**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
PS C:\Program Files\Lens> .\Lens.exe
PS C:\Program Files\Lens>
info: ­ƒôƒ Setting Lens as protocol client for lens://
info: ­ƒôƒ succeeded Ô£à
```

**Kubeconfig:**
Quite often the problems are caused by malformed kubeconfig which the application tries to load. Please share your kubeconfig, remember to remove any secret and sensitive information.
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJT...JTm9ZRElrUjkyWWoybGpKNm12eUNheWJ2VT0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
    server: https://925334FBF88268531E9A9EC20021795F.sk1.eu-west-1.eks.amazonaws.com
  name: SVTS-K8s-K8s-SVTS-Cluster
- cluster:
    certificate-authority-data: LS0tLS1CRUdJ...NBVEUtLS0tLQo=
    server: https://0979ECE0F56E8A969DFD36153BDF74FD.yl4.eu-west-1.eks.amazonaws.com
  name: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJ...lGSUNBVEUtLS0tLQo=
    server: https://kubernetes.docker.internal:6443
  name: docker-desktop
- cluster:
    certificate-authority-data: LS0tLS1CRUdJ...0VSVElGSUNBVEUtLS0tLQo=
    server: https://34.76.197.49
  name: gke_zcs-smarty-cloud_europe-west1_gke-smarty-cloud
- cluster:
    certificate-authority: C:\Users\esche\.minikube\ca.crt
    server: https://172.17.244.185:8443
  name: minikube
contexts:
- context:
    cluster: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
    namespace: id-smsa-svts-platform-namespace
    user: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
  name: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-for-desktop
- context:
    cluster: gke_zcs-smarty-cloud_europe-west1_gke-smarty-cloud
    user: gke_zcs-smarty-cloud_europe-west1_gke-smarty-cloud
  name: gke_zcs-smarty-cloud_europe-west1_gke-smarty-cloud
- context:
    cluster: SVTS-K8s-K8s-SVTS-Cluster
    namespace: identity-smsa
    user: devops-infocert-test
  name: infocert-test
- context:
    cluster: minikube
    user: minikube
  name: minikube
current-context: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
kind: Config
preferences: {}
users:
- name: arn:aws:eks:eu-west-1:956251722582:cluster/SVTS-Factory-EKS-Cluster
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region
      - eu-west-1
      - eks
      - get-token
      - --cluster-name
      - SVTS-Factory-EKS-Cluster
      command: aws
      env:
      - name: AWS_PROFILE
        value: ic-test
- name: devops-infocert-test
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region
      - eu-west-1
      - eks
      - get-token
      - --cluster-name
      - SVTS-K8s-K8s-SVTS-Cluster
      command: aws
      env:
      - name: AWS_PROFILE
        value: ictst
- name: docker-desktop
  user:
    client-certificate-data: LS0tLS...LS0K
    client-key-data: LS0tL...VktLS0tLQo=
- name: gke_zcs-smarty-cloud_europe-west1_gke-smarty-cloud
  user:
    auth-provider:
      config:
        access-token: ya2...GIRiN0
        cmd-args: config config-helper --format=json
        cmd-path: C:\Users\esche\AppData\Local\Google\Cloud SDK\google-cloud-sdk\bin\gcloud.cmd
        expiry: "2021-04-23T14:04:04Z"
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
      name: gcp
- name: minikube
  user:
    client-certificate: C:\Users\esche\.minikube\client.crt
    client-key: C:\Users\esche\.minikube\client.key

```

**Additional context**
Nothing more to add.

### LINKED ISSUE #3363: Lens does not show Workloads even after full uninstall
**Describe the bug**
After installing Lens, the workspace does not show anything.  Any attempts to expand Workloads, Overview, Apps, etc do not show expected items.

I have attempted to uninstall Lens, remove the Lens directory and cache as well as completely remove the kubeconfig file but the issue persists.  I installed from k8lens.dev and using `brew`, but the problem persisted.  Several other users do not have this issue. 

I am able to perform all actions via `kubectl` so I do not think this is a permissions issue within the cluster.

**To Reproduce**
1. Install Lens
2. use `gcloud` to populate the kubeconfig to connect to GKE cluster
3. Open Lens
4. Notice that nothing shows on Lens

**Expected behavior**
Expected behavior is to open Lens and be able to see all workloads, nodes and other objects. 

**Screenshots**

**Environment (please complete the following information):**
- Lens Version: 5.0.2
- OS: Mac Catalina 10/15/7
- Installation method (e.g. snap or AppImage in Linux): Website or `brew`

**Logs:**
Only error in logs i see is:
```
error: Can't find working API for the Kubernetes resource ingresses
```

**Additional context**
I have tried multiple versions 4.x and 5.x without success.

### LINKED ISSUE #4493: Cannot view workload after upgrading to Lens 5.3.2-latest.20211201.1 with restricted permissions
**Describe the bug**
After upgrading Lens from 5.2.x the Workloads section and most of the resources in the sidebar no longer load. We had allowed permissions so that users could list namespaces specifically so that they did not have to input the list of accessible namespaces manually.

**To Reproduce**
- No full read access on the cluster
- Permissions to get/list/watch on all namespaces
- Full permissions on workloads in a specific namespace

**Expected behavior**
I am expecting to be able to see resources in namespaces that I have access to, like in previous Lens versions

**Screenshots**
Without specifying an accessible namespace in the cluster settings. Notice how almost all the resources are not loaded on the sidebar:

image

After specifying an accessible namespace in the cluster settings (this is the behavior that we were seeing in a previous version of Lens without having to specify the namespace), all the resources to which we have access are now back as it was prior to the update :

image

**Environment (please complete the following information):**
- Lens Version: 5.3.2-latest.20211201.1
- OS: Windows 10
- Installation method (e.g. snap or AppImage in Linux): installed from executable on Lens website, updated to latest through auto-update mechanism.

**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
There are no relevant logs (no errors/warning) that show any kind of exception we would be encountering.
With developer tools, there are also no errors/warnings, everything seems fine.
```

**Kubeconfig:**
This is not Kubeconfig related as this was working prior to upgrading Lens.

**Additional context**
We had somewhat figured out the required RBAC through trial and error in previous versions as there is no clear Lens documentation on that topic. We do not want to give everyone full read access to the cluster.

### LINKED ISSUE #6151: Cluster permissions with narrow namespace coverage
**Describe the bug**

Closes https://github.com/lensapp/lens/issues/1700
Closes https://github.com/lensapp/lens/issues/4493

This has been a long standing short coming of our current design where we only check the first 10 namespaces for list permissions for all the built in resources. This works for small clusters and for users that are cluster admins (can do everything) but for users that only have a very narrow set of permissions this basically makes Lens unusable.

**To Reproduce**
1. Have a cluster with a large number of namespaces
2. Have a user only have list permissions in a namespace that was not one of the first 10 created
3. Try and use lens

**Expected behavior**
Lens should still be usable within those namespaces

### LINKED ISSUE #6613: Menu elements missing on cluster with many namespaces
**Describe the bug**
Most of the menu entries are missing/hidden.

I looked into the code and only the first 10 namespaces are used to determine the allowed ressources.
https://github.com/lensapp/lens/blob/v6.2.0/src/common/cluster/cluster.ts#L681

**To Reproduce**
- connect to cluster with more then 10 namespaces
- do not grant permissions for the first 10 namespaces
- menu items are hidden

**Expected behavior**
- menu items are shown when permissions are set for any namespace

**Screenshots**
image

**Environment (please complete the following information):**
- Lens Version: v6.2.0
- OS: OSX 12.6

### PR DESCRIPTION
**Pull request by ahippler**
fixes #6613 

the SelfSubjectRulesReview lists all resources for a namespaces. This reduces the api calls from `resource_count * namespace_count` to `namespace_count`

### CODE REVIEW

**Comment by Nokel81** [comment]
Very cool indeed. I wonder what the performance penalties are with clusters that have hundreds or thousands of namespaces. Maybe we should also (as future work) add the ability to disable this check and just show all the menus.

Also it seems like `getAllowedResources` is called every 30s which seems a bit overkill. Would you be able to make it so that it only refreshed every 15minutes like metadata is?

File: src/common/cluster/authorization-namespace-review.injectable.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+import type { KubeConfig } from "@kubernetes/client-node";
+import { AuthorizationV1Api } from "@kubernetes/client-node";
+import logger from "../logger";
```
Comment (Nokel81) [inline src/common/cluster/authorization-namespace-review.injectable.ts]: logger shouldn't be imported globally in new code, you can get the instance from `di.inject(loggerInejctable)` within the `instantiate` function below.

File: src/common/cluster/authorization-namespace-review.injectable.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { getInjectable } from "@ogre-tools/injectable";
+
+export type NamespaceResources = (namespace: string) => Promise<string[]>;
+
+/**
+ * @param proxyConfig This config's `currentContext` field must be set, and will be used as the target cluster
+   */
+export function authorizationNamespaceReview(proxyConfig: KubeConfig): NamespaceResources {
```
Comment (Nokel81) [inline src/common/cluster/authorization-namespace-review.injectable.ts]: 1. This really shouldn't be exported
2. This should probably be a builder of sorts which first takes `dependencies: Dependencies` and then returns this current function. That way you can inject `logger` in.

```ts
export type RequestNamespaceResources = (namespace: string) => Promise<string[]>;
export type AuthorizationNamespaceReview = (proxyConfig: KubeConfig) => RequestNamespaceResources;

interface Dependencies {
  logger: Logger;
}

const authorizationNamespaceReview( = ({
  logger,
}: Dependencies): AuthorizationNamespaceReview => {
  ...
};
```

**Comment by Nokel81** [comment]
fixes https://github.com/lensapp/lens/issues/2609
fixes https://github.com/lensapp/lens/issues/6151
fixes https://github.com/lensapp/lens/issues/1700
fixes https://github.com/lensapp/lens/issues/4493
fixes https://github.com/lensapp/lens/issues/3363
fixes https://github.com/lensapp/lens/issues/2399

File: src/common/cluster/authorization-namespace-review.injectable.ts (hunk 1 above)
Comment (ahippler) [reply src/common/cluster/authorization-namespace-review.injectable.ts]: this was copied from src/common/cluster/authorization-review.injectable.ts

I changed it in both files

**Review by Nokel81 — DISMISSED** [review]
Thanks very much for this! This is a great first step. As I stated above we might want to have an option to disable this in the future for people if they have too big of a cluster

**Comment by ahippler** [comment]
fixed the linter issues

**Review by Nokel81 — DISMISSED** [review]

**Comment by ahippler** [comment]
SelfSubjectRulesReview returns `*` if all resources of one (more more) group(s) have the same permissions. Its is not possible to list the allowed resources for `*` without a list of supported resources. So i ended up collecting the available resources with some other requests

File: integration/__tests__/cluster-pages.tests.ts [hunk 3]
```diff
@@ -417,7 +411,7 @@ const scenarios = [
   {
     expectedSelector: "h5.title",
     parentSidebarItemTestId: "sidebar-item-link-for-user-management",
-    sidebarItemTestId: "sidebar-item-link-for-pod-security-policies",
```
Comment (ahippler) [inline integration/__tests__/cluster-pages.tests.ts:420]: minikube does not have PodSecurityPolicy

**Comment by Nokel81** [comment]
Please resolve conflicts

**Review by Nokel81 — APPROVED** [review]

File: integration/__tests__/cluster-pages.tests.ts (hunk 3 above)
Comment (Nokel81) [reply integration/__tests__/cluster-pages.tests.ts:420]: Then why did this test pass previously?

Comment (ahippler) [reply integration/__tests__/cluster-pages.tests.ts:420]: looks like the api reports missing ressources as allowed:

```
$ kubectl --context minikube auth can-i list podsecuritypolicies
Warning: the server doesn't have a resource type 'podsecuritypolicies'

yes
```

**Pull request merged**